src/ott/geometry/geodesic.py

```diff
@@ -0,0 +1,308 @@
+
+    Args:
+      G: Adjacency matrix.
+      t: Constant used when approximating the geodesic exponential kernel.
+        If `None`, use :math:`\frac{1}{|E|} \sum_{(u, v) \in E} weight(u, v)`
+        :cite:`crane:13`. In this case, the ``graph`` must be specified
+        and the edge weights are all assumed to be positive.
+      directed: Whether the ``graph`` is directed. If not, it will be made
+        undirected as :math:`G + G^T`. This parameter is ignored when  directly
+        passing the Laplacian, which is assumed to be symmetric.
+      normalize: Whether to normalize the Laplacian as
+        :math:`L^{sym} = \left(D^+\right)^{\frac{1}{2}} L
+        \left(D^+\right)^{\frac{1}{2}}`, where :math:`L` is the
+        non-normalized Laplacian and :math:`D` is the degree matrix.
+      kwargs: Keyword arguments for :class:`~ott.geometry.graph.Graph`.
+
+    Returns:
+      The graph geometry.
+    """
+    assert G.shape[0] == G.shape[1], G.shape
+
+    if directed:
+      G = G + G.T
+
+    degree = jnp.sum(G, axis=1)
+    laplacian = jnp.diag(degree) - G
+
+    if normalize:
+      inv_sqrt_deg = jnp.diag(
+          jnp.where(degree > 0.0, 1.0 / jnp.sqrt(degree), 0.0)
+      )
+      laplacian = inv_sqrt_deg @ laplacian @ inv_sqrt_deg
+
+    if t is None:
+      t = (jnp.sum(G) / jnp.sum(G > 0.)) ** 2
+
+    return cls(laplacian, t=t, **kwargs)
+
+  def apply_kernel(
+      self,
+      scaling: jnp.ndarray,
+      eps: Optional[float] = None,
+      axis: int = 0,
+  ) -> jnp.ndarray:
+    r"""Apply :attr:`kernel_matrix` on positive scaling vector.
+
+    Args:
+    scaling: Scaling to apply the kernel to.
+    eps: passed for consistency, not used yet.
+    axis: passed for consistency, not used yet.
+
+    Returns:
+    Kernel applied to ``scaling``.
+    """
+
+    def compute_laplacian(adjacency_matrix: jnp.ndarray) -> jnp.ndarray:
+      """Compute the Laplacian matrix from the adjacency matrix.
+
+      Args:
+          adjacency_matrix: An (n, n) array representing the
+          adjacency matrix of a graph.
+
+      Returns:
+          An (n, n) array representing the Laplacian matrix.
+      """
+      degree_matrix = jnp.diag(jnp.sum(adjacency_matrix, axis=0))
+      return degree_matrix - adjacency_matrix
+
+    def compute_largest_eigenvalue(laplacian_matrix, k):
+      """Compute the largest eigenvalue of the Laplacian matrix.
+
+      Args:
+          laplacian_matrix: An (n, n) array representing the Laplacian matrix.
+          k: Number of eigenvalues/vectors to compute.
+
+      Returns:
+          The largest eigenvalue of the Laplacian matrix.
+      """
+      n = laplacian_matrix.shape[0]
+      initial_directions = jax.random.normal(jax.random.PRNGKey(0), (n, k))
+      # Convert the Laplacian matrix to a dense array
+      #laplacian_array = laplacian_matrix.toarray()
+      eigvals, _, _ = lobpcg_standard(laplacian_matrix, initial_directions, m=k)
```

Comment:
`lobpcg_standard` doesn't seem to accept a sparse matrix by default, but you can define the matrix-vector product using as `jax.experimental.sparse.sparsify(lambda v: lap @ v)`. Currently the best supported format is the [BCOO](https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.sparse.BCOO.html). When refactoring, please start assuming that the laplacian can also be a BCOO sparse matrix.
Here are some further pointers: https://github.com/google/lineax/issues/24

Reply:
Done.